Warn if kernel does not support overlay with selinux

rhvgoyal opened the pull request:
We first added error to not allow overlay with selinux enabled. Then later
we removed it as kernel was getting close to get the support. But this 
means user does not get meaningful message on old kernels.

This patch reintrouces error back. Difference is that it dynamically
tries to detect if underlying kernel supports overlayfs with selinux
or not. And if it does not, it errors out.

It will not error out if it detects that kernel supports overlayfs with
selinux.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
cc @rhatdan

rhvgoyal commented:
cc @thaJeztah

thaJeztah commented:
Thanks! @justincormack @dmcgowan ptal as well

justincormack commented:
This does not seem a good idea, that symbol could well disappear in a kernel release tomorrow and this will break.

rhatdan commented:
If a kernel does not have this symbol then it probably does not support SELinux and Overlay, so it should fail.  I guess this could cause a problem if the kernel did support SELinux but changed the symbol name.

rhvgoyal commented:
@justincormack sure, this symbol can be renamed in kernel. In that case we will change docker as well so that it works with new kernel. Old kernels will continue to work anyway with existing code. So it is just a matter of keeping up with new kernel and we will be just fine.

rhvgoyal commented:
I think it is better than not giving error message to user at all.

rhvgoyal commented:
Once we are well past the state where oldest kernel docker supports has overlayfs selinux support, we can get rid of that code. That might be few years. Till then it will provide good protection against wrong configuration.

rhvgoyal commented:
Fixed lint errors.

thaJeztah commented:
but if a user updates the kernel, and docker breaks, that's pretty bad

rhvgoyal commented:
@thaJeztah We releases fixes all the time. And what are the chances that will happen. Unfortunately we don't have a clean API to detect which features kernel supports. Now we have only two options.

- Let user run docker on a kernel where feature is not supported without error. User will get all kind of odd errors and take a while to figure out this is not right configuration.

- Put a mechanism to error out when configuration is wrong. (With a very small chance that mechanism is broken over kernel upgrade and will require fixing).

I personally prefer second option. If kernel changes the symbol, that will most likely be detected much early in kernel -rc releases and docker will be fixed. IOW, chances that symbol changed and breakage remained undetected and kernel went to user without a fix in docker are very small.

IMHO pros outweigh cons in this case.

vdemeester commented:
I think it make sense to at least warn the user (or error out) instead of saying nothing.

justincormack commented:
We don't release fixes all the time, eg we will have users on 1.11.2 Docker for quite some time, but we are not going to patch it for changes in kernel symbols in future kernel releases, it is only going to get security updates. Kernel symbols are not a supported stable interface. This might work for a Linux distro where you ship the kernel and Docker at the same time with patches, but we have users with years worth of different kernels and Docker versions, and there is absolutely no guarantee this will not break as soon as the 4.9 kernel.

rhvgoyal commented:
@justincormack Agreed that it will be hard to release fixes for released/stable docker versions for new kernel. 

How about giving a warning instead of error message. That way downside is that already released docker is emitting a wrong warning if kernel symbol name changes.

rhvgoyal committed:
Warn if kernel does not support overlay/overlay2 with selinux

We first added error to not allow overlay with selinux enabled. Then later
we removed it as kernel was getting close to get the support. But this 
means user does not get meaningful message on old kernels.

This patch introduces a warning (Instead of error). Difference is that it
dynamically tries to detect if underlying kernel supports overlayfs with
selinux or not. And if it does not, it warns.

It will not warn if it detects that kernel supports overlayfs with selinux.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
Refreshed patch. Changed error to warning.

LK4D4 commented:
LGTM
ping @justincormack

vdemeester reviewed (approved):
LGTM 🐸

vdemeester commented:
re-ping @justincormack 🙇

AkihiroSuda reviewed (approved):
LGTM

PTAL @justincormack

AkihiroSuda commented:
@justincormack Can we merge this PR? Now this is just warn and hence does not error out.

LK4D4 commented:
LGTM

kolyshkin commented:
Here is a quote from RHEL7.4 release notes (https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/7.4_release_notes/new_features_file_systems)

> CHAPTER 9. FILE SYSTEMS
> SELinux security labels are now supported on the OverlayFS file system
> With this update, the OverlayFS file system now supports SELinux security labels. When using Docker containers with the OverlayFS storage driver, you no longer have to configure Docker to disable SELinux support for the containers. (BZ#1297929)

Bugzilla link: https://bugzilla.redhat.com/show_bug.cgi?id=1297929
